uvicorn/main.py

```diff
@@ -388,53 +393,162 @@ def main(
+    run(
+        app,
+        host=host,
+        port=port,
+        uds=uds,
+        fd=fd,
+        loop=loop,
+        http=http,
+        ws=ws,
+        ws_max_size=ws_max_size,
+        ws_ping_interval=ws_ping_interval,
+        ws_ping_timeout=ws_ping_timeout,
+        ws_per_message_deflate=ws_per_message_deflate,
+        lifespan=lifespan,
+        env_file=env_file,
+        log_config=LOGGING_CONFIG if log_config is None else log_config,
+        log_level=log_level,
+        access_log=access_log,
+        interface=interface,
+        debug=debug,
+        reload=reload,
+        reload_dirs=reload_dirs if reload_dirs else None,
+        reload_includes=reload_includes if reload_includes else None,
+        reload_excludes=reload_excludes if reload_excludes else None,
+        reload_delay=reload_delay,
+        workers=workers,
+        proxy_headers=proxy_headers,
+        server_header=server_header,
+        date_header=date_header,
+        forwarded_allow_ips=forwarded_allow_ips,
+        root_path=root_path,
+        limit_concurrency=limit_concurrency,
+        backlog=backlog,
+        limit_max_requests=limit_max_requests,
+        timeout_keep_alive=timeout_keep_alive,
+        ssl_keyfile=ssl_keyfile,
+        ssl_certfile=ssl_certfile,
+        ssl_keyfile_password=ssl_keyfile_password,
+        ssl_version=ssl_version,
+        ssl_cert_reqs=ssl_cert_reqs,
+        ssl_ca_certs=ssl_ca_certs,
+        ssl_ciphers=ssl_ciphers,
+        headers=[header.split(":", 1) for header in headers],
+        use_colors=use_colors,
+        factory=factory,
+        app_dir=app_dir,
+    )
+
+
+@typing.overload  # type: ignore[misc]
+def run(
+    app: typing.Union[ASGIApplication, str],
+    *,
+    host: str,
+    port: int,
+    uds: str,
+    fd: int,
+    loop: LoopSetupType,
+    http: HTTPProtocolType,
+    ws: WSProtocolType,
+    ws_max_size: int,
+    ws_ping_interval: float,
+    ws_ping_timeout: float,
+    ws_per_message_deflate: bool,
+    lifespan: LifespanType,
+    interface: InterfaceType,
+    debug: bool,
+    reload: bool,
+    reload_dirs: typing.Optional[typing.List[str]],
+    reload_includes: typing.Optional[typing.List[str]],
+    reload_excludes: typing.Optional[typing.List[str]],
+    reload_delay: float,
+    workers: int,
+    env_file: str,
+    log_config: str,
+    log_level: str,
+    access_log: bool,
+    proxy_headers: bool,
+    server_header: bool,
+    date_header: bool,
+    forwarded_allow_ips: str,
+    root_path: str,
+    limit_concurrency: int,
+    backlog: int,
+    limit_max_requests: int,
+    timeout_keep_alive: int,
+    ssl_keyfile: str,
+    ssl_certfile: str,
+    ssl_keyfile_password: str,
+    ssl_version: int,
+    ssl_cert_reqs: int,
+    ssl_ca_certs: str,
+    ssl_ciphers: str,
+    headers: typing.Sequence[typing.Sequence[str]],
+    use_colors: bool,
+    app_dir: str,
+    factory: bool,
+) -> None:
+    """Run uvicorn via Python code.
+
+    Args:
```

Comment:
is the docstring necessary here or the type hints are enough to get the completion ?

Reply:
Annotation is enough for that, the diverting are a plus